dev/buildtool/git.py

```diff
@@ -461,6 +461,41 @@ def to_yaml(self, with_commit_messages=True):
 class GitRunner(object):
   """Helper class for interacting with Git"""
 
+  @staticmethod
+  def __normalize_repo_url(url):
+    """Normalize a repo url for purposes of checking equality.
+
+    Returns:
+      Either a tuple (HOST, OWNER, PATH) if url is a github-like URL
+         assumed to be in the form <PROTOCOL://HOST/OWNER/PATH> where
+         or ssh@<HOST>:<USER><REPO>
+         in these cases, a '.git' REPO postfix is considered superfluous.
+
+      Otherwise a string assuming the url is a local path
```

Comment:
How does this behave if you try to compare a local path and remote? Does this occur in practice? Seems like those two would be incomparable using this helper.

Reply:
They are different repos unless you were somehow looking at a local path that the "remote" was serving from. A clone of a repo isnt the same as the repo it cloned for this purpose.

This function is a safety check to minimize risk of pushing upstream when the upstream is the origin but with a different url (e.g. ssh vs https or https vs http). In the case of local vs remote, safety is already assured.

FWIW, the GitRemoteRepository class has the origin and the upstream so tracks the relationship you are alluding to. There is a function in this module that can derive one of those objects from an existing repository if the fetch stage were in a separate process so that relationship isnt locally known.